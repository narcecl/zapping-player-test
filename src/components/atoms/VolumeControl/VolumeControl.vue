<script setup lang="ts">
import { ref } from 'vue';
import ActionButton from '@/components/atoms/ActionButton/ActionButton.vue';
import VolumeIcon from '@/components/icons/VolumeIcon.vue';
import { usePlayerStore, useLanguageStore } from '@/stores';
import { storeToRefs } from 'pinia';
import MuteIcon from '@/components/icons/MuteIcon.vue';

const barRef = ref<HTMLElement | null>(null);
let isDragging = false;

const { $t } = useLanguageStore();
const playerStore = usePlayerStore();
const { volume, visibilityStatus } = storeToRefs(playerStore);

const ToggleVolumeControl = () => {
    playerStore.toggleVisibility('volume', !visibilityStatus.value.volume);
};

const setVolumeFromEvent = (event: MouseEvent) => {
    if (!barRef.value) return;
    const rect = barRef.value.getBoundingClientRect();
    const x = event.clientX - rect.left;
    const newVolume = Math.max(0, Math.min(1, x / rect.width));
    playerStore.changeVolume(newVolume);
};

const handleClick = (event: MouseEvent) => {
    setVolumeFromEvent(event);
};

const startDrag = (event: MouseEvent) => {
    isDragging = true;
    setVolumeFromEvent(event);
    window.addEventListener('mousemove', handleDrag);
    window.addEventListener('mouseup', stopDrag);
};

const handleDrag = (event: MouseEvent) => {
    if (isDragging) setVolumeFromEvent(event);
};

const stopDrag = () => {
    isDragging = false;
    window.removeEventListener('mousemove', handleDrag);
    window.removeEventListener('mouseup', stopDrag);
};
</script>

<template>
    <div class="volume__control flex items-center gap-4">
        <ActionButton @click="ToggleVolumeControl" :aria-label="$t('controls_volume')">
            <!-- I used other sound icon for UX purposes -->
            <VolumeIcon v-if="volume > 0.1" aria-hidden="true" class="text-white" />
            <MuteIcon v-if="!volume" aria-hidden="true" class="text-white" />
        </ActionButton>

        <Transition name="slide-left">
            <div
                v-if="visibilityStatus.volume"
                class="volume__control__bar"
                ref="barRef"
                @mousedown="startDrag"
                @click="handleClick"
            >
                <div
                    class="volume__control__bar__progress"
                    :style="{ width: `${volume * 100}%` }"
                />
            </div>
        </Transition>
    </div>
</template>

<style lang="scss" scoped>
.volume__control {
    min-width: 300px;

    &__bar {
        width: 192px;
        height: 8px;
        position: relative;
        background: rgba(white, 0.15);
        border-radius: 4px;
        overflow: hidden;
        cursor: pointer;

        &__progress {
            background: var(--color-brand-primary);
            height: 100%;
        }
    }
}

.slide-left {
    &-enter-active,
    &-leave-active {
        transition:
            opacity 0.3s ease,
            transform 0.2s ease;
    }
    &-enter-from {
        opacity: 0;
        transform: translateX(-10px);
    }
    &-leave-to {
        opacity: 0;
        transform: translateX(-10px);
    }
}
</style>
