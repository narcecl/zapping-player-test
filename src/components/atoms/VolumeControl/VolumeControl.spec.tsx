import { describe, test, expect, vi } from 'vitest';

import { mount } from '@vue/test-utils';
import VolumeControl from './VolumeControl.vue';
import { createTestingPinia } from '@pinia/testing';

describe('Badge', () => {
    test('Default render (button)', () => {
        const wrapper = mount(<VolumeControl>F</VolumeControl>, {
            global: {
                plugins: [
                    createTestingPinia({
                        createSpy: vi.fn,
                    }),
                ],
            },
        });
        expect(wrapper.html()).toMatchSnapshot();

        const icon = wrapper.get('svg');
        expect(icon).toBeDefined();
        expect([...wrapper.element.classList]).includes('volume__control');
    });
});
