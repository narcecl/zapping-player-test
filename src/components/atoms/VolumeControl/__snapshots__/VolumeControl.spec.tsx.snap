// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Badge > Default render (button) 1`] = `
"<div data-v-8208447f="" class="volume__control flex items-center gap-4"><button data-v-8208447f="" class="action_button">
    <!--v-if--><svg data-v-8208447f="" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" aria-hidden="true" class="text-white">
      <g fill="none" stroke="currentColor" stroke-width="2">
        <path d="M3.158 13.93a3.75 3.75 0 0 1 0-3.86a1.5 1.5 0 0 1 .993-.7l1.693-.339a.45.45 0 0 0 .258-.153L8.17 6.395c1.182-1.42 1.774-2.129 2.301-1.938S11 5.572 11 7.42v9.162c0 1.847 0 2.77-.528 2.962c-.527.19-1.119-.519-2.301-1.938L6.1 15.122a.45.45 0 0 0-.257-.153L4.15 14.63a1.5 1.5 0 0 1-.993-.7Z"></path>
        <path stroke-linecap="round" d="m15 15l6-6m0 6l-6-6"></path>
      </g>
    </svg>
  </button>
  <transition-stub data-v-8208447f="" name="slide-left" appear="false" persisted="false" css="true">
    <!--v-if-->
  </transition-stub>
</div>"
`;
